child_process.flushStdio: resume _consuming streams

davidvgalbraith opened the pull request:
Hey! I'm new here. Tried to follow all the guidelines. Thanks for your attention!

When a client calls read() with a nonzero argument
on a Socket, that Socket sets this._consuming to true.
It never sets this._consuming back to false.
child_process.flushStdio currently doesn't flush
any streams where _consuming is truthy. But that means
that it never flushes any stream that has ever been read from.
This prevents a child process from ever closing if one of
its streams has been read from, causing issue #4049.

child_process.flushStdio should flush streams even if their
_consuming is set to true. Then it will close even after a read.

Fishrock123 commented:
cc @nodejs/streams

Fishrock123 commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/881/

cjihrig commented on test/sequential/test-child-process-flush-stdio.js:
```diff
@@ -0,0 +1,13 @@
+const cp = require('child_process');
+const common = require('../common');
+
+var p = cp.spawn('yes');
```
Is `yes` a valid Windows command?

calvinmetcalf replied:
I had my coworker on windows type yes into the prompt and it is not a command

Fishrock123 replied:
Can also confirm it is not.

cjihrig commented:
Is there a reason the test needs to be in `sequential` instead of `parallel`?

cjihrig commented:
A couple of questions about the test, but mostly LGTM. I would like sign off from a streams person if possible.

davidvgalbraith commented:
I moved the test into `parallel`, and I changed its use of `yes` to the cross-platform `echo`, double-checking that it still fails on master and passes on this branch.

cjihrig commented on test/parallel/test-child-process-flush-stdio.js:
```diff
@@ -0,0 +1,13 @@
+const cp = require('child_process');
+const common = require('../common');
+
+var p = cp.spawn('echo');
```
Can you please make this `const`.

cjihrig commented on test/parallel/test-child-process-flush-stdio.js:
```diff
@@ -0,0 +1,13 @@
+
+var p = cp.spawn('echo');
+
+p.on('close', common.mustCall(function() {}));
```
Maybe add an assertion that the `close` arguments are what is expected.

davidvgalbraith commented:
I made `p` a `const`, and I added validating assertions to the `close` event handler.

cjihrig commented:
Started a CI run: https://ci.nodejs.org/job/node-test-pull-request/883/

The linter doesn't like your indentation - https://ci.nodejs.org/job/node-test-linter/434/console

Fishrock123 commented:
@cjihrig note: currently the CI is private. See https://github.com/nodejs/node/issues/4029#issuecomment-160610575

davidvgalbraith commented:
How embarrassing. I've indented the requisite two spaces. Forsooth I can't see anything at the CI link.

Fishrock123 commented:
@davidvgalbraith could you run `make jslint`? That should do the trick. :)

davidvgalbraith commented:
`make jshint` is looking green as of https://github.com/davidvgalbraith/node/commit/127a82ecc950ac73199ded537db5b1161b022855.

cjihrig commented:
@Fishrock123 good catch about the CI being private.

@davidvgalbraith the CI looks fine. There were a couple failures unrelated to this change. LGTM

cjihrig commented:
Thanks! Landed in 34b535f4caba366789d949211e4369b1e1d01152

jasnell commented:
@cjihrig ... would you consider this a bug fix (and thus suitable for LTS) or an add?

cjihrig commented:
I would consider this a bug fix.

silverwind commented:
Looks like this new test here is failing on CentOS:

https://ci.nodejs.org/job/node-test-commit-linux/1387/nodes=centos5-32/tapResults/
https://ci.nodejs.org/job/node-test-commit-linux/1387/nodes=centos5-64/tapResults/
```
	test-child-process-flush-stdio.js	
	# Mismatched &lt;anonymous&gt; function calls. Expected 1, actual 0.
	# at Object.&lt;anonymous&gt; (/home/iojs/build/workspace/node-test-commit-linux/nodes/centos5-64/test/parallel/test-child-process-flush-stdio.js:8:22)
	# at Module._compile (module.js:399:26)
	# at Object.Module._extensions..js (module.js:406:10)
	# at Module.load (module.js:345:32)
	# at Function.Module._load (module.js:302:12)
	# at Function.Module.runMain (module.js:431:10)
	# at startup (node.js:138:18)
	# at node.js:976:3
```

cjihrig commented:
Yes. See https://github.com/nodejs/node/issues/4125

TheAlphaNerd commented:
Not going to land this until we are ready to land #4215 into LTS... I think it needs a bit more time
